### ISSUE TITLE
`single-match` makes incorrect suggestion for unsafe code

### ISSUE THREAD

**Issue by nyurik** [rust-lang/rust-clippy#10808]
### Summary

This type of code creates incorrect suggestions for all lints like `single-match` and `single-match-else`:

```rust
match bar {
    Some(v) => unsafe { ... }
    ...
}
```

### Reproducer

Several non-passing tests for this issue have already been created in #10809

Run `cargo clippy -- -W clippy::single-match` with this code:

```rust
fn main() {
    foo(None);
}

pub fn foo(bar: Option<u32>) {
    match bar {
        Some(v) => unsafe {
            let r = &v as *const u32;
            println!("{}", *r);
        },
        _ => {}
    }
}
```

I expected Clippy to suggest this code:

```rust
    if let Some(v) = bar {
        unsafe {
            let r = &v as *const u32;
            println!("{}", *r);
        }
    }
```

Instead, it suggested this broken code:

```rust
    if let Some(v) = bar unsafe {
        let r = &v as *const u32;
        println!("{}", *r);
    }
```

### Version

```text
rustc 1.69.0 (84c898d65 2023-04-16)
binary: rustc
commit-hash: 84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc
commit-date: 2023-04-16
host: x86_64-unknown-linux-gnu
release: 1.69.0
LLVM version: 15.0.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by nyurik** [comment]
I added a few failing tests in #10809
